### PR TITLE
Initial sui-edge-proxy crate

### PR DESCRIPTION
**Pull request by johnjmartin**
## Description 

Adds an initial version of the sui-edge-proxy crate for use proxying edge requests for JSON RPC fullnodes. 

## Test plan 

Tested in our testnet and mainnet dedicated benchmark infra

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: crates/sui-edge-proxy/src/handlers.rs [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            return Ok(Response::builder()
+                .status(StatusCode::INTERNAL_SERVER_ERROR)
+                .body(Body::from("Failed to read request body"))
+                .unwrap());
+        }
+    };
+
+    match parts
+        .headers
+        .get("Client-Request-Method")
```
Comment (mystenmark) [inline crates/sui-edge-proxy/src/handlers.rs:75]: this isn't a standard http header afact, so we should name it `X-Sui-Client-Request-Method` or something, right?

File: crates/sui-edge-proxy/src/handlers.rs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+            "application/json",
+        ))
+    {
+        info!("Content type is not application/json");
+        return Ok(Response::builder()
+            .status(StatusCode::BAD_REQUEST)
+            .body(Body::from("Content type must be application/json"))
+            .unwrap());
+    }
+    let body_bytes = match axum::body::to_bytes(body, usize::MAX).await {
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-edge-proxy/src/handlers.rs:62]: also I think we need a lower limit than usize::MAX, or else anyone can OOM the server, right?

File: crates/sui-edge-proxy/src/handlers.rs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            }
+        }
+    }
+}
+
+async fn proxy_request(
+    state: AppState,
+    parts: Parts,
+    body_bytes: Bytes,
+    use_execution_peer: bool,
```
Comment (mystenmark) [inline crates/sui-edge-proxy/src/handlers.rs:108]: Let's use an enum here:

```
enum PeerRole {
   Execution,
   Read,
}
```

File: crates/sui-edge-proxy/src/handlers.rs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        .observe(body_bytes.len() as f64);
+
+    let peer_config = if use_execution_peer {
+        &state.execution_peer
+    } else {
+        &state.read_peer
+    };
+
+    let target_url = peer_config.address.clone();
+    // remove host header to avoid interfering with reqwest auto-host header
```
Comment (mystenmark) [inline crates/sui-edge-proxy/src/handlers.rs:132]: can you expand on this? ordinarily a proxy would preserve the host header set by the client

File: crates/sui-edge-proxy/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    let (_guard, _filter_handle) = telemetry_subscribers::TelemetryConfig::new()
+        .with_env()
+        .with_prom_registry(&prometheus_registry)
+        .init();
+
+    info!("Metrics server started at {}", config.metrics_address);
+
+    // Build a reqwest client that supports HTTP/2
+    let client = reqwest::ClientBuilder::new()
+        .http2_prior_knowledge()
```
Comment (mystenmark) [inline crates/sui-edge-proxy/src/main.rs:48]: if a node operator has for some reason disabled http2, will they get a useful error?

File: crates/sui-edge-proxy/src/config.rs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+use url::Url;
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct ProxyConfig {
+    pub listen_address: SocketAddr,
+    pub metrics_address: SocketAddr,
+    pub execution_peer: PeerConfig,
+    pub read_peer: PeerConfig,
+    pub max_idle_connections: Option<usize>,
```
Comment (mystenmark) [inline crates/sui-edge-proxy/src/config.rs:16]: can you document the implications of the max_idle_connections param?

File: crates/sui-edge-proxy/src/handlers.rs (hunk 1 above)
Comment (johnjmartin) [reply crates/sui-edge-proxy/src/handlers.rs:75]: is `X-` commonly used for custom headers? I copied the format of other custom headers in the sdk: https://github.com/MystenLabs/sui/blob/main/sdk/typescript/src/client/http-transport.ts#L95-L98

**Review by mystenmark — APPROVED** [review]

**Pull request merged**